File: ssafy-study/week1/BOJ_11725_bfs.java
import java.io.*;
import java.util.*;

public class BOJ_11725_bfs {

  static List<Integer>[] adj;
  static int[] p;
  static int N;

  static void bfs() {
    Queue<Integer> q = new LinkedList<>();
    q.add(1);
  
    while (!q.isEmpty()) {
      int cur = q.poll();
      for (int nxt : adj[cur]) {
        if (p[nxt] != 0) continue;
        q.add(nxt);
        p[nxt] = cur;
      }
    }
  }

  static void init() {
    adj = new ArrayList[N + 1];
    for (int i = 1; i < N + 1; ++i) adj[i] = new ArrayList<>();
    p = new int[N + 1];
  }

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    
    // 입력
    N = Integer.parseInt(br.readLine());
    init();

    for (int i = 0; i < N - 1; ++i) {
      StringTokenizer st = new StringTokenizer(br.readLine());

      int u = Integer.parseInt(st.nextToken());
      int v = Integer.parseInt(st.nextToken());
      adj[u].add(v);
      adj[v].add(u);
    }

    // bfs
    bfs();

    // 출력
    StringBuilder sb = new StringBuilder();
    for (int i = 2; i <= N; ++i) sb.append(p[i]).append('\n');
    System.out.print(sb);
  }
}

File: ssafy-study/week1/BOJ_11725_bfs.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> adj[100001];
int p[100001];

void bfs() {
  queue<int> q;
  q.push(1);

  while (!q.empty()) {
    int cur = q.front(); q.pop();
    for (int nxt : adj[cur]) {
      if (p[nxt]) continue; // 이미 부모 노드가 정해졌으면 continue
      q.push(nxt);
      p[nxt] = cur;
    }
  }
}

int main(void) {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int N;
  cin >> N;

  for (int i = 0; i < N - 1; ++i) {
    int u, v;
    cin >> u >> v;
    adj[u].push_back(v);
    adj[v].push_back(u);
  }
  bfs();
  for (int i = 2; i <= N; ++i) cout << p[i] << '\n';
}
